lib/SILOptimizer/Mandatory/TFDeabstraction.cpp

```diff
@@ -365,43 +373,69 @@ static BuiltinInst *simplifyOperands(BuiltinInst *inst, TFDeabstraction &TFDA) {
   // Predicate that returns true if the specified type is an address type for
   // a loadable (non-address-only) value.
   auto isLoadableAddressType = [&](SILType type) -> bool {
-    return type.isAddress() && type.isLoadable(inst->getModule());
+    return type.isAddress() && type.isLoadable(origInst->getModule());
   };
 
   // Predicate that returns true if an operand of the specified type should be
   // rewritten - either to load an address argument or expand a struct
   // parameter.
-  auto canSimplifyOperand = [&](SILType type) -> bool {
+  auto canSimplifyOperand = [&](SILType type,
+                                SILTensorOpInfo::OperandClass operandClass)
+      -> bool {
     return isLoadableAddressType(type) ||
-           getPrimitiveStructField(type.getASTType()) != nullptr;
+           getPrimitiveStructField(type.getASTType()) != nullptr ||
+           operandClass == SILTensorOpInfo::OperandClass::Out;
   };
 
   // If we don't have to change any operands, don't rewrite the builtin.
   bool mustChangeBuiltin = false;
-  for (auto &op : inst->getAllOperands()) {
-    if (canSimplifyOperand(op.get()->getType())) {
+  for (auto operandAndClass : zip(origInst->getAllOperands(),
+                                  opInfo.operandClasses)) {
+    auto operand = std::get<0>(operandAndClass).get();
+    auto operandClass = std::get<1>(operandAndClass);
+    if (canSimplifyOperand(operand->getType(), operandClass.second)) {
       mustChangeBuiltin = true;
       break;
     }
   }
 
-  if (!mustChangeBuiltin) return inst;
+  if (!mustChangeBuiltin) return origInst;
 
   // Mark the function as being mutated.
   TFDA.aboutToChangeFunction();
 
   // Okay, we do have to simplify something.  Scan through and rewrite operands.
-  SILBuilder B(inst);
+  SILBuilder B(origInst);
   SmallVector<SILValue, 8> operands;
-  for (auto &op : inst->getAllOperands()) {
-    auto operand = op.get();
+  SmallVector<std::pair<StringRef, SILTensorOpInfo::OperandClass>, 4>
+      operandClasses;
+  llvm::Optional<SILValue> outParameterAddress;
+  for (auto operandAndClass : zip(origInst->getAllOperands(),
+                                  opInfo.operandClasses)) {
+    auto operand = std::get<0>(operandAndClass).get();
+    auto operandClass = std::get<1>(operandAndClass);
+
+    // If this is an out parameter, take note of it (so that we can deal with it
+    // later), and exclude it from the list of rewritten operands.
+    if (operandClass.second == SILTensorOpInfo::OperandClass::Out) {
+      assert(operand->getType().isAddress());
+      if (!operand->getType().isLoadable(origInst->getModule())) {
```

Comment:
It's better to make this an assertion because generic specialization is guaranteed to have taken place at this stage.